### ISSUE TITLE
How to use node values for roughness param in TidalFlowCalculator

### ISSUE THREAD

**Issue by ifoxfoot** [landlab/landlab#1567]
Hello, I'm having trouble figuring out how to use a field with roughness values (stored in nodes) for the roughness parameter in `TidalFlowCalculator`. I keep getting the error `ValueError: Array passed to function decorated with use_field_name_array_or_value is not the size of fields at link`. See the code example below

```
from landlab import RasterModelGrid
from landlab.components import TidalFlowCalculator
from landlab.plot import imshow
from pylab import show, figure

#make raster data
grid = RasterModelGrid((3, 5), xy_spacing=2.0) 
z = grid.add_zeros("topographic__elevation", at = "node")
z[:] = -50.0  # mean water depth is 50 m below MSL
rough = grid.add_zeros("roughness", at = "node")
rough[:] = 0.02 # mean roughness...

#plot raster
figure('Elevations from the field')  # new fig, with a name
imshow.imshow_grid_at_node(grid, "topographic__elevation")
show()

#plot raster
figure('Roughness vals')  # new fig, with a name
imshow.imshow_grid_at_node(grid, "roughness")
show()

tfc = TidalFlowCalculator(grid, tidal_range=2.0, tidal_period=4.0e4, roughness=rough)

```
I'm not sure if this is a bug or if roughness is supposed to be stored in links? Or I'm doing something wrong? I'm new to landlab so any help would be appreciated! 

- Using landlab version 2.6.0 
- Python 3.11.0 | packaged by conda-forge | (main, Jan 15 2023, 05:44:48) [Clang 14.0.6 ] on darwin
- Using MacOS Monterey 12.6.3

**Comment by gregtucker** [comment]
If roughness is a field rather than a scalar, it needs to be defined at links (I realize that's not at all clear in the documentation...)

**Comment by mdpiper** [comment]
@ifoxfoot @gregtucker If you can suggest some language to clarify the docs, I can make a quick update.

**Comment by gregtucker** [comment]
I noticed that @mcflugen follows the practice of listing field-like parameters like:

`u : (n_links,) ndarray`

In this case, we could use something like:

`roughness : float or (n_links,) ndarray`

@ifoxfoot would that help or does that seem too opaque?

**Comment by ifoxfoot** [comment]
@gregtucker and @mdpiper, okay the error is making a lot more sense now! And yes, I think that any mention of links in the parameter description would be helpful.

So to convert a raster of roughness values to links could I use the `calc_grad_at_link` function?

**Comment by mcflugen** [comment]
> I noticed that @mcflugen follows the practice of listing field-like parameters like:
>
> u : (n_links,) ndarray

I did this following *scipy*'s example but, if it's not clear, we could do something else.

> So to convert a raster of roughness values to links could I use the calc_grad_at_link function?

If you just need to have roughness at links, maybe just this following?

```python
rough = grid.add_full("roughness", 0.2, at="link")
```

**Comment by mcflugen** [comment]
@ifoxfoot As a follow-up on my last comment.

As @gregtucker mentioned, if you are using a constant value for roughness you can just pass that along to the `TidalFlowCalculator`,
```python
tfc = TidalFlowCalculator(grid, tidal_range=2.0, tidal_period=4.0e4, roughness=0.2)
```
If you want roughness values at both nodes and links, one option would be to do what you suggested and map roughness from nodes to links by taking the gradient.
```python
grid.at_link["roughness"] = grid.calc_grad_at_link("roughness")
```
Another way would be to use one of the node-to-link mappers. For example,
```python
from numpy.random import uniform

grid.at_node["roughness"] = uniform(0.0, 0.2, size=grid.number_of_nodes)
grid.at_link["roughness"] = grid.map_mean_of_link_nodes_to_link("roughness")
```

**Comment by ifoxfoot** [comment]
@mcflugen okay that's super helpful! Yes my raster of roughness values is not constant, sorry I should have clarified that!

**Comment by gregtucker** [comment]
I would just add that `calc_grad_at_link` would not be a good choice if you want to map roughness values defined on nodes to links, because it applies the gradient rather than the original value (so for example if your roughness were uniform, the gradient would be zero everywhere). A better option is  `map_mean_of_link_nodes_to_link` or a related mapper function (there are also functions that take the maximum or minimum value at the two nodes).

**Comment by ifoxfoot** [comment]
Right, good point. I will have to think on which one will work best. I think the `map_max_of_link_nodes_to_link` might be best for my case because I have some lonely high roughness pixels and I don't want to dilute them by taking the mean. Thanks again for all your help!
